lane.py

```diff
@@ -800,11 +807,34 @@ def custom_lists_for_identifier(self, list_data_source, list_identifier):
             list_data_source_id = list_data_source.id
         else:
             list_data_source_id = None
+
+        list_featured_works_query = None
         if lists:
             list_ids = [x.id for x in lists]
+            list_featured_works_query = self.extract_list_featured_works_query(lists)
         else:
             list_ids = None
-        return list_data_source_id, list_ids
+        return list_data_source_id, list_ids, list_featured_works_query
+
+    def extract_list_featured_works_query(self, lists):
+        if not lists:
+            return None
+        if isinstance(lists[0], int):
+            # We have CustomList ids instead of CustomList objects.
+            lists = self._db.query(CustomList).filter(CustomList.id.in_(lists))
+            lists = lists.all()
+
+        lists = [custom_list for custom_list in lists if custom_list.featured_works]
+        if not lists:
+            return None
+
+        work_ids = list()
+        for custom_list in lists:
+            work_ids += [work.id for work in custom_list.featured_works]
+
+        works_query = self._db.query(Work).with_labels().\
+            filter(Work.id.in_(work_ids))
+        return works_query
```

Comment:
I don't know if this will be a problem anytime soon but there are a couple things that would be a lot ore efficient here.

First, you could do this in one query joining CustomListEntry against LicensePool against Work.

Or, you could do this in a couple queries by gathering all the custom list IDs, then defining a class method on CustomList that takes a list of list IDs and looks for the featured works in any of those lists, all at once.

I think this is okay for now, but if we start having really big lists this is an easy place to save some time.